paddle/fluid/distributed/auto_parallel/spmd_rules/reshape_spmd_rule.cc

```diff
@@ -0,0 +1,205 @@
+      tgt_idx++;
+    }
+
+    // deal with the singleton case
+    if (s == 1 && t != 1) {
+      // case [1] [a]
+      tgt_idx--;
+      tgt_splitted_shape.clear();
+    } else if (s != 1 && t == 1) {
+      src_idx--;
+      src_dims.clear();
+    } else {
+      while (s != t) {
+        if (s < t) {
+          src_dims.emplace_back(src_idx);
+          s *= src_shape[src_idx];
+          src_idx++;
+        } else {
+          tgt_splitted_shape.emplace_back(inferred_tgt_shape[tgt_idx]);
+          t *= inferred_tgt_shape[tgt_idx];
+          tgt_idx++;
+        }
+      }
+    }
+
+    if (tgt_splitted_shape.size() > 0) {
+      std::vector<DimTrans*> input_dims;
+      for (int64_t i = 0, n = src_dims.size(); i < n; i++) {
+        int64_t in_dim = src_dims[i];
+        if (src_shape[in_dim] > 1) {
+          input_dims.emplace_back(new InputDim(in_dim));
+        }
+      }
+      DimTrans* flatten = make_flatten(input_dims);
+
+      for (int64_t i = 0, n = tgt_splitted_shape.size(); i < n; i++) {
+        ret.emplace_back(make_split(flatten, tgt_splitted_shape, i));
+      }
+    }
+  }
+  return ret;
+}
+
+std::pair<std::vector<TensorDistAttr>, std::vector<TensorDistAttr>>
+paddle::distributed::auto_parallel::ReshapeSPMDRule::InferForward(
+    const std::vector<DistTensorSpec>& input_specs,
+    const paddle::framework::AttributeMap& attrs) {
+  // step0: Verify Input Args Based on Reshape Logic
+  int64_t ninputs = input_specs.size();
+  PADDLE_ENFORCE_EQ(
+      ninputs,
+      1,
+      phi::errors::InvalidArgument("The size of InputSpec in reshape must "
+                                   "be equal to 1, but got [%d].",
+                                   ninputs));
+  VerifySpecs(input_specs, "reshape");
+
+  // step1: build the transformation from
+  // original shape to target shape
+  std::vector<int64_t> src_shape = input_specs[0].shape();
+  std::vector<int64_t> tgt_shape =
+      ExtractAttr<std::vector<int64_t>>("shape", attrs);
+
+  // handle the '0' values in target shape, '0' indicates
+  // that the target shape is equal to the source shape
+  for (int64_t i = 0, n = tgt_shape.size(); i < n; i++) {
+    if (tgt_shape[i] == 0) {
+      tgt_shape[i] = src_shape[i];
+    }
+  }
+
+  std::vector<DimTrans*> trans = MakeReshapeDimTrans(src_shape, tgt_shape);
+
+  // step2: infer the dims mapping of input (if reshard is
+  // needed) and output from the dimension transformation.
+  std::vector<std::vector<int64_t>> dims_mapping_vec =
+      InferFromDimTrans(input_specs[0], trans);
+
+  // step3: update the dist attributes of input
+  // and output with the inferred dims mapping
+  TensorDistAttr new_input_dist_attr(input_specs[0].dist_attr());
+  new_input_dist_attr.set_dims_mapping(dims_mapping_vec[0]);
+  TensorDistAttr output_dist_attr(input_specs[0].dist_attr());
+  output_dist_attr.set_dims_mapping(dims_mapping_vec[1]);
+
+  VLOG(4) << "Reshape: input_shape: [" << str_join(src_shape)
```

Comment:
TODO: think about print useful info about tensor axes and dims_mapping for debug:

idea1: construct einsum notation for debug and giving corresponding axes between input and output a specific character, therefore user could be notified that those axes are related.

idea2: print out the DimTrans and make the info readable.

Reply:
Done, print the transformation info.